src/libcrun/kontain_nonkmexec.c

```diff
@@ -0,0 +1,347 @@
+         continue;
+      }
+      char* s = strchr(buf, '\n');
+      if (s != NULL) {
+         *s = 0;
+      }
+      // Split into fields
+      if (split_into_fields(buf, fields) != 0) {
+         rc = EINVAL;
+         libcrun_fail_with_error(rc, "Couldn't split <%s> into fields", buf);
+      }
+
+      epp = calloc(sizeof(execpath_entry_t), 1);
+      if (epp == NULL) {
+         libcrun_fail_with_error(ENOMEM, "Couldn't allocate %d bytes of memory", sizeof(execpath_entry_t));
+      }
+
+      // Compile regular expression.
+      rc = regcomp(&epp->re, fields[0], REG_EXTENDED | REG_NOSUB);
+      if (rc != 0) {
+         char regerrbuf[128];
+         regerror(rc, &epp->re, regerrbuf, sizeof(regerrbuf));
+         libcrun_fail_with_error(rc, "Couldn't compile regular expression %s, %s", fields[0], regerrbuf);
+      }
+      epp->path = strdup(fields[1]);
+      if (epp->path == NULL) {
+         rc = ENOMEM;
+         libcrun_fail_with_error(rc, "Couldn't strdup %s", fields[1]);
+      }
+
+      // Compute hash of target file
+      rc = sha256_file(fields[1], epp->sha256);
+      if (rc != 0) {
+         libcrun_fail_with_error(rc, "Couldn't compute hash of file %s", fields[1]);
+      }
+      // Does the hash match what the exec_ok file has?
+      if (strcmp(epp->sha256, fields[2]) != 0) {
+         printf("%s: Computed hash %s doesn't match expected hash %s\n",
+                fields[1],
+                epp->sha256,
+                fields[2]);
+         libcrun_fail_with_error(EILSEQ, "hash on file %s differ: expected: %s got: %s", epp->path, fields[2], epp->sha256);
+      }
+
+      // Chain entry on to the end of the list
+      if (tail == NULL) {
+         SLIST_INSERT_HEAD(&statep->execpath_head, epp, link);
+      } else {
+         SLIST_INSERT_AFTER(tail, epp, link);
+      }
+      tail = epp;
+      epp = NULL;
+   }
+   fclose(f);
+
+   // Remember the file's mod time
+   struct stat statb;
+   rc = stat(statep->execpath_dbpath, &statb);
+   if (rc == 0) {
+      statep->execpath_mtime = statb.st_mtim;
+   } else {
+      libcrun_fail_with_error(errno, "stat %s failed", statep->execpath_dbpath);
+   }
+}
+
+// Find an entry in the execpath_ok db that matches execpath.
+static int
+kontain_execpath_lookup(execpath_state_t* statep, const char* execpath, execpath_entry_t** eppp)
+{
+   int rc;
+   struct stat statb;
+
+   if (statep->execpath_dbpath == NULL) {
+      *eppp = NULL;
+      return 0;
+   }
+   if (stat(statep->execpath_dbpath, &statb) != 0) {
+      if (errno != ENOENT) {
+         libcrun_fail_with_error(errno, "can't access allowed nonkm executable list %s", statep->execpath_dbpath);
+         return errno;
+      }
+      // No allowed executable file, so no executable is allowed to run without km
+      *eppp = NULL;
+      return 0;
+   }
+   if (memcmp(&statb.st_mtim, &statep->execpath_mtime, sizeof(struct timespec)) != 0) {
+      kontain_execpath_builddb(statep);
```

Comment:
it feels excessive to build the list on every search.  I guess the assumption is that there is one request (search) for one invocation of crun ? then it's OK , but then the list head being non-null on enterying search in really an assert - we NEVER expect this to happen, right ?

Reply:
We only rebuild if the config file has changed since last read.
Maybe there needs to be a better name than execpath_mtime.

Reply:
that's where I am confused. How it could be called read twice ? It is one invokation per exec(), isn't it ?

Reply:
Yes, there currently is a single invocation.